File: Codility/Lesson/0011.Sieve-of-Eratosthenes/CountSemiprimes/CountSemiprimes.go
package countsemiprimes

import "fmt"

// TODO:
func Solution(N int, P []int, Q []int) []int {
	prime := make([]int, N+1)
	i := 2
	for i*i <= N {
		if prime[i] == 0 {
			k := i * i
			for k <= N {
				if prime[k] == 0 {
					prime[k] = i
				}
				k = k + i
			}
		}
		i++
	}
	fmt.Println("prime: ", prime)

	// 當前一共有幾個半質數
	semiprime := make([]int, N+1)
	for i := 1; i < len(prime); i++ {
		p := prime[i]
		if p == 0 {
			semiprime[i] = semiprime[i-1]
			continue
		}
		if prime[i/p] == 0 {
			semiprime[i] = semiprime[i-1] + 1
		} else {
			semiprime[i] = semiprime[i-1]
		}
	}
	fmt.Println("semiprime: ", semiprime)

	result := make([]int, len(P))
	for i, p := range P {
		result[i] = semiprime[Q[i]] - semiprime[p-1]
	}
	return result
}

// TODO: Wrong answer. Need check again
func Solution2(N int, P []int, Q []int) []int {
	semiPrime := []int{}

	// 半質數:兩個質數的乘積所得的自然數我們稱之為半質數.
	// 4, 6, 9, 10, 14, 15, 21,22,25,26,33,34,35,38,39,46,49,51,55,57,58,62,65,69,74,77,82,85,86,87,91,93,94,95,106, ...
	// 它們包含1及自己在內合共有3或4個因數
	for i := 1; i <= N; i++ {
		factorCount := 0
		sign := 0
		for j := 1; j*j <= i; j++ {
			if i%j == 0 {
				factorCount++
				f := i / j
				if f != j {
					if f == j*j {
						// 3個相同: ex i=27, j=3, f=9
						sign = 1
						semiPrime = append(semiPrime, 0)
						break
					} else {
						factorCount++
					}
				}
			}
			if factorCount > 4 {
				sign = 1
				semiPrime = append(semiPrime, 0)
			}
		}
		if sign != 1 {
			if factorCount >= 3 {
				semiPrime = append(semiPrime, i)
			} else {
				semiPrime = append(semiPrime, 0)
			}
		}
	}
	fmt.Println("semiPrime", semiPrime)

	// 當前array和前面一共有幾個半質數
	indexMap := make(map[int]int)
	// 如果是半質數添加到 map
	semiMap := make(map[int]int)
	count := 0
	for i := 0; i < len(semiPrime); i++ {
		if semiPrime[i] != 0 {
			count++
			indexMap[semiPrime[i]] = count
			semiMap[semiPrime[i]] = 0
		} else {
			indexMap[i+1] = count
		}
	}
	fmt.Println("indexMap : ", indexMap)
	fmt.Println("semiMap : ", semiMap)

	result := []int{}
	for i := 0; i < len(P); i++ {
		if _, ok := semiMap[P[i]]; ok {
			result = append(result, indexMap[Q[i]]-indexMap[P[i-1]])
		} else {
			result = append(result, indexMap[Q[i]]-indexMap[P[i]])
		}
	}
	return result
}

File: Codility/Lesson/0011.Sieve-of-Eratosthenes/CountSemiprimes/CountSemiprimes.py
# https://github.com/Anfany/Codility-Lessons-By-Python3/blob/master/L11_Sieve%20of%20Eratosthenes/11.2%20CountSemiprimes.md
def solution(N, P, Q):
    """
    返回由数组P、Q的元素组成的区间内，不大于N的半素数的个数, 时间复杂度O(N * log(log(N)) + M)
    :param N: 半素数的最大值
    :param P: 数组
    :param Q: 数组
    :return: 每次查询,得到的半素数的个数
    """
    #  半素数只有3或4个因子，并且不能是素数的立方，例如(1, 3, 9, 27)(1, 5, 25, 125)这种情况
    #  首先计算出不大于N的半素数列表，是半素数的为其值，不是的为0
    semi_prime = []
    k =0
    for i in range(1, N + 1):
        factor_count = 0
        sign = 0
        for j in range(1, int(i ** 0.5) + 1):
            if i % j == 0:
                factor_count += 1
                f = i / j
                if f != j:
                    if f == j ** 2:
                        sign = 1
                        semi_prime.append(0)
                        break
                    else:
                        factor_count += 1
            if factor_count > 4:
                sign = 1
                semi_prime.append(0)
                break
        if sign != 1:
            if factor_count >= 3:
                semi_prime.append(i)
            else:
                semi_prime.append(0)

    index_dict = {}  # 得出当前数值以及前面一共有几个半素数
    semi_dict = {}  # 如果是半素数，则添加到字典中
    count = 0
    for index, value in enumerate(semi_prime):
        if value != 0:
            count += 1
            index_dict[value] = count
            semi_dict[value] = 0
        else:
            index_dict[index + 1] = count
    # index_dict {1: 0, 2: 0, 3: 0, 4: 1, 5: 1, 6: 2, 7: 2, 8: 2, 9: 3, 10: 4, 11: 4, 12: 4, 13: 4, 14: 5, 15: 6, 16: 6, 17: 6, 18: 6, 19: 6, 20: 6, 21: 7, 22: 8, 23: 8, 24: 8, 25: 9, 26: 10}
    #semi_dict {4: 0, 6: 0, 9: 0, 10: 0, 14: 0, 15: 0, 21: 0, 22: 0, 25: 0, 26: 0}
    print("index_dict",index_dict)
    print("semi_dict",semi_dict)

    result_list = []  # 开始计算，在指定区间内有几个半素数
    for i, j in zip(P, Q):
        if i in semi_dict:
            result_list.append(index_dict[j] - index_dict[i] + 1)
        else:
            result_list.append(index_dict[j] - index_dict[i])

    return result_list


if __name__ == '__main__':
    solution(26,[1, 4, 16],[26, 10, 20])